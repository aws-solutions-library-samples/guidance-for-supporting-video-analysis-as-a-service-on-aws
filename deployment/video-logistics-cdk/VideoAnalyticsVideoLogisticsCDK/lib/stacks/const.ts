export const FORWARDING_RULES_BUCKET_NAME = "videoanalytics-fwd-rules";
export const FORWARDING_RULES_ACCESS_LOGS_BUCKET_NAME =
  "videoanalytics-fwd-rules-access-logs";
export const FORWARDING_RULES_SQS_QUEUE_NAME =
  "aws-ip-videoanalytics-VLControlPlane-forwardingRulesQueue";
export const FORWARDING_RULES_TABLE_NAME = "ForwardingRulesTable";
export const FORWARDING_RULES_PK_NAME = "CustomerAccountId";
export const FORWARDING_RULES_SK_NAME = "ForwardingRulesName";
export const VIDEO_EXPORT_JOB_TABLE_NAME = "VideoExportJobTable";
export const VIDEO_EXPORT_JOB_RESULT_TABLE_NAME = "VideoExportResultTable";
export const VIDEO_EXPORT_SFN_STATE_MACHINE_NAME = "VideoExportStateMachine";
export const VIDEO_EXPORT_JOB_GSI_NAME = "DeviceJobStatus-Index";
export const VIDEO_EXPORT_JOB_GSI_PK = "CustomerAccountId_DeviceId";
export const VIDEO_EXPORT_JOB_GSI_SK = "JobStatus";
export const JOB_ID = "JobId";
export const EXPORT_FILE_START_TIME = "ExportFileStartTime";
export const VIDEO_EXPORT_JOB_PARTITION_KEY = "CustomerAccountId_DeviceId";
export const JOB_STATUS = "JobStatus";
export const VIDEO_EXPORT_RESULTS_GSI_1ST_NAME = "JobIDStatus-Index";
export const VIDEO_EXPORT_RESULTS_GSI_1ST_SK = "SegmentJobStatus";
export const VIDEO_EXPORT_RESULTS_GSI_2ND_NAME = "JobIDEndTime-Index";
export const VIDEO_EXPORT_RESULTS_GSI_2ND_SK = "ExportFileEndTime";
export const VIDEO_EXPORT_ACCOUNT_JOB_ID_GSI = "AccountIDJobID-Index";
export const VIDEO_EXPORT_ACCOUNT_JOB_ID_GSI_PK = "CustomerAccountId";
// Step function state parameters
export const EXECUTION_COUNTER = "$.Counter";
export const EXPORT_RESULT_PATH = "$.exportResult";
export const EXECUTION_RESULT_IN_STATUS = "$.executionResult";
export const UPDATED_TIMESTAMP =
  "$.exportResult.Payload.newLastProcessedTimestamp";
export const IS_COMPLETE_CONDITION = "$.exportResult.Payload.isJobComplete";
export const LAST_PROCESSED_TIMESTAMP = "$.lastProcessedTimestamp";
export const WAIT_TIME_BETWEEN_LOOP = "$.exportResult.Payload.waitTime";
export const FAILED_DYNAMIC_CAUSE = "$.Cause";
// used for when execution Lambda failed cause forwarding after statusUpdate
export const FAILED_FORWARD_CAUSE = "$.Payload.Cause";
export const SHOULD_STREAM_FROM_DEVICE_PATH = "$.shouldStreamFromDevice";
export const PARTITION_KEY_PATH = "$.partitionKey";
export const SORT_KEY_PATH = "$.sortKey";
export const RESULT_PATH_ERROR = "$.error";
export const ERROR_MESSAGE_PATH = "$.error.Cause";
export const RESULT_PATH = "$.output";
export const enum PLAYBACK_SESSION_STATUS {
  FAILED = "FAILED",
  COMPLETED = "COMPLETED",
}
export const RAW_VIDEO_TIMELINE_TABLE_NAME = "RawVideoTimelineTable";
export const VIDEO_TIMELINE_TABLE_NAME = "VideoTimelineTable";
export const RAW_VIDEO_TIMELINE_PK_NAME = "CustomerId_DeviceId";
export const VIDEO_TIMELINE_PK_NAME = "CustomerId_DeviceId_TimeUnit";
export const VIDEO_TIMELINE_SK_NAME = "UnitTimestamp";
export const VIDEO_TIMELINE_TTL_ATTRIBUTE_NAME = "ExpirationTimestamp";
export const RAW_VIDEO_TIMELINE_SORT_KEY_NAME = "Timestamp";
export const TIMELINE_BUCKET_NAME = "videoanalytics-timeline-bucket";
// TODO: Update handler to match new lambda handler path
export const DENSITY_UPDATE_LAMBDA_HANDLER_PATH =
  "com.amazon.awsvideoanalyticsvlcontrolplane.timeline.VideoDensityUpdateLambda::handleRequest";
export const EXPORT_LAMBDA_HANDLER_PATH =
  "com.amazon.awsvideoanalyticsvlcontrolplane.timeline.VideoTimelineS3ExportLambda::handleRequest";
export const LAMBDA_SERVICE_PRINCIPAL = 'lambda.amazonaws.com';
export const LAMBDA_MANAGED_POLICY_NAME = 'service-role/AWSLambdaBasicExecutionRole';
// TODO: Update handler to match new lambda handler path
export const TIMELINE_FORWARDER_HANDLER_PATH =
  'com.amazon.awsvideoanalyticsvlcontrolplane.timeline.TimelineForwarderLambda::handleRequest';
export const LAMBDA_PACKAGE_NAME = 'AWSVideoAnalyticsVLControlPlane-1.0';
  "com.amazon.awsvideoanalyticsvlcontrolplane.timeline.TimelineForwarderLambda::handleRequest";

export const VL_ACTIVITY_JAVA_PATH_PREFIX =
  "com.amazonaws.videoanalytics.videologistics.activity";
export const VL_WORKFLOW_JAVA_PATH_PREFIX =
  "com.amazonaws.videoanalytics.videologistics.workflow";
export const LAMBDA_ASSET_PATH =
  "../../../assets/lambda-built/video-logistics-assets/VideoAnalyticsVideoLogisticsControlPlane-1.0-beta.jar";
export const OPEN_API_SPEC_PATH = "../../../assets/model/video-logistics/openapi-conversion/openapi/VideoAnalytic.openapi.json";