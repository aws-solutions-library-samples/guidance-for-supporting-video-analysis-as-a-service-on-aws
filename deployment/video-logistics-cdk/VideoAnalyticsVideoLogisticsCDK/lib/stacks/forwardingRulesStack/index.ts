// export * from './forwardingRulesStack';
export * from './opensearchStack';