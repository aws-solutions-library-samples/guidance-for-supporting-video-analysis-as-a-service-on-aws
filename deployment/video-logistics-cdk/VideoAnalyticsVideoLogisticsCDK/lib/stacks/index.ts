export * from './forwardingRulesStack';
export * from './schedulerStack';
export * from './timelineStack';
export * from './videoExportStack';
export * from './workflowStack';
export * from './serviceStack';
