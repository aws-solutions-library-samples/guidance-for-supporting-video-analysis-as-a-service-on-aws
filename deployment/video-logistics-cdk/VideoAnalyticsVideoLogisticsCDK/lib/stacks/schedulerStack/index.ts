export * from './openSearchPitCreation';
export * from './schedulerStack';