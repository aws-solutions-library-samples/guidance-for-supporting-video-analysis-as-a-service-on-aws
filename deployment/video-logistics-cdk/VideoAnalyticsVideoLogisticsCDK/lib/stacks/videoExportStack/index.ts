export * from './videoExportWorkflowResources';
export * from './videpExportStack';