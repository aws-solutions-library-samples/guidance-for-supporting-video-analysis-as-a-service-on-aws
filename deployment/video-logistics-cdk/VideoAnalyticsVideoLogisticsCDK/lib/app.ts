import { App } from "aws-cdk-lib";
import { 
    // ForwardingRulesStack,
    ServiceStack,
    TimelineStack,
    VideoExportStack,
    WorkflowStack
 } from "./stacks";
import { AWSRegion } from "video_analytics_common_construct";
import { VideoLogisticsBootstrapStack } from "./stacks/bootstrapStack/videoLogisticsBootstrapStack";
import { BulkInferenceStack } from "./stacks/workflowStack/bulkInferenceStack";
import { OpenSearchStack } from "./stacks/forwardingRulesStack";

const app = new App();

// Get the account and region from environment variables or use defaults
const account =
  process.env.CDK_DEPLOY_ACCOUNT ||
  process.env.CDK_DEFAULT_ACCOUNT ||
  "YOUR_DEFAULT_ACCOUNT";
const region = (process.env.CDK_DEPLOY_REGION ||
  process.env.CDK_DEFAULT_REGION ||
  "us-west-2") as AWSRegion;

// Common environment config
const envConfig = {
    env: {
        account: account,
        region: region
    },
    region: region,
    account: account
};

// Create all stacks
// new ForwardingRulesStack(app, 'VideoLogisticsForwardingRulesStack', envConfig);
// new SchedulerStack(app, 'VideoLogisticsSchedulerStack', envConfig);
// new TimelineStack(app, 'VideoLogisticsTimelineStack', envConfig);
// new VideoExportStack(app, 'VideoLogisticsVideoExportStack', envConfig);
new VideoLogisticsBootstrapStack(app, 'VideoLogisticsBootstrapStack', envConfig);
let opensearchStack = new OpenSearchStack(app, 'VideoLogisticsOpensearchStack', envConfig);
new BulkInferenceStack(app, 'VideoLogisticsBulkInferenceStack', {...envConfig, opensearchEndpoint: opensearchStack.opensearchEndpoint});
new WorkflowStack(app, 'VideoLogisticsWorkflowStack', envConfig);
new ServiceStack(app, 'VideoLogisticsServiceStack', envConfig);

app.synth();
