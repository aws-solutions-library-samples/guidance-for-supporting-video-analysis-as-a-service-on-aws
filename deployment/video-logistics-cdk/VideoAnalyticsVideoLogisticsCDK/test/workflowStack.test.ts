import { createStackSnapshotTest } from "./snapshotTestUtil";
import { WorkflowStack } from "../lib/stacks";

createStackSnapshotTest("WorkflowStack", WorkflowStack);
