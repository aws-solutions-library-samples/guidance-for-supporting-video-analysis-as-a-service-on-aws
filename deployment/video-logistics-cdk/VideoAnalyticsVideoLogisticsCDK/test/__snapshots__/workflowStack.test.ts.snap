// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkflowStack Snapshot test 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "RegisterDeviceWorkflowDataForwarderLambdaLogGroup2C7FBF4F": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "LogGroupName": "KvsCreateLambdaLogGroup",
        "RetentionInDays": 3653,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "RegisterDeviceWorkflowFailAndCleanupFvlRegHandlerLogGroupFB8AF6F7": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "LogGroupName": "FailAndCleanupFvlRegHandlerLogGroup",
        "RetentionInDays": 3653,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "RegisterDeviceWorkflowFailAndCleanupFvlRegHandlerRole331A94CC": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "dynamodb:GetItem",
                    "dynamodb:PutItem",
                  ],
                  "Effect": "Allow",
                  "Resource": [
                    {
                      "Fn::Join": [
                        "",
                        [
                          "arn:",
                          {
                            "Ref": "AWS::Partition",
                          },
                          ":dynamodb:us-east-1:123456789012:table/VLRegisterDeviceJobTable",
                        ],
                      ],
                    },
                    {
                      "Fn::GetAtt": [
                        "VLRegisterDeviceJobTableA904F593",
                        "Arn",
                      ],
                    },
                  ],
                },
                {
                  "Action": "sts:AssumeRole",
                  "Effect": "Allow",
                  "Resource": [
                    {
                      "Fn::Join": [
                        "",
                        [
                          "arn:",
                          {
                            "Ref": "AWS::Partition",
                          },
                          ":iam::*:role/CrossAccountRoleForStartFVLRegisterDevice",
                        ],
                      ],
                    },
                    "arn:aws:iam::*:role/CrossAccountRoleForStartFVLRegisterDevice-iad",
                  ],
                },
                {
                  "Action": [
                    "kinesisvideo:DeleteStream",
                    "kinesisvideo:DeleteSignalingChannel",
                  ],
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":kinesisvideo:us-east-1:123456789012:*",
                      ],
                    ],
                  },
                },
                {
                  "Action": [
                    "kms:Decrypt",
                    "kms:Encrypt",
                    "kms:ReEncrypt*",
                    "kms:GenerateDataKey*",
                  ],
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":kms:us-east-1:123456789012:key/*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "AccessPolicy",
          },
        ],
        "RoleName": "FailAndCleanupFvlRegHandlerRole",
      },
      "Type": "AWS::IAM::Role",
    },
    "RegisterDeviceWorkflowFailAndCleanupVlRegHandler41AB87C8": {
      "DependsOn": [
        "RegisterDeviceWorkflowFailAndCleanupFvlRegHandlerRole331A94CC",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-123456789012-us-east-1",
          "S3Key": "32d918bda097c43ad86cd13b3d0279caa352c738afc045d18826e367a85f990d.jar",
        },
        "Description": "Lambda responsible for cleaning up KVS resources and updating the failure status in ddb.",
        "Environment": {
          "Variables": {
            "airportCode": "iad",
            "tableName": "VLRegisterDeviceJobTable",
          },
        },
        "Handler": "com.amazonaws.videoanalytics.videologistics.workflow.FailAndCleanupVLDeviceRegistrationHandler::handleRequest",
        "LoggingConfig": {
          "LogGroup": {
            "Ref": "RegisterDeviceWorkflowFailAndCleanupFvlRegHandlerLogGroupFB8AF6F7",
          },
        },
        "MemorySize": 2048,
        "Role": {
          "Fn::GetAtt": [
            "RegisterDeviceWorkflowFailAndCleanupFvlRegHandlerRole331A94CC",
            "Arn",
          ],
        },
        "Runtime": "java17",
        "Timeout": 720,
      },
      "Type": "AWS::Lambda::Function",
    },
    "RegisterDeviceWorkflowKVSStreamCreatorLambdaRole92F0C304": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "dynamodb:GetRecords",
                    "dynamodb:GetItem",
                    "dynamodb:Query",
                    "dynamodb:PutItem",
                    "dynamodb:UpdateItem",
                  ],
                  "Effect": "Allow",
                  "Resource": [
                    {
                      "Fn::Join": [
                        "",
                        [
                          "arn:",
                          {
                            "Ref": "AWS::Partition",
                          },
                          ":dynamodb:us-east-1:123456789012:table/VLRegisterDeviceJobTable",
                        ],
                      ],
                    },
                    {
                      "Fn::GetAtt": [
                        "VLRegisterDeviceJobTableA904F593",
                        "Arn",
                      ],
                    },
                  ],
                },
                {
                  "Action": "sts:AssumeRole",
                  "Effect": "Allow",
                  "Resource": [
                    {
                      "Fn::Join": [
                        "",
                        [
                          "arn:",
                          {
                            "Ref": "AWS::Partition",
                          },
                          ":iam::*:role/CrossAccountRoleForStartFVLRegisterDevice",
                        ],
                      ],
                    },
                    "arn:aws:iam::*:role/CrossAccountRoleForStartFVLRegisterDevice-iad",
                  ],
                },
                {
                  "Action": [
                    "kinesisvideo:CreateStream",
                    "kinesisvideo:CreateSignalingChannel",
                  ],
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":kinesisvideo:us-east-1:123456789012:*",
                      ],
                    ],
                  },
                },
                {
                  "Action": [
                    "kms:Decrypt",
                    "kms:Encrypt",
                    "kms:ReEncrypt*",
                    "kms:GenerateDataKey*",
                  ],
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":kms:us-east-1:123456789012:key/*",
                      ],
                    ],
                  },
                },
                {
                  "Action": [
                    "kms:Decrypt",
                    "kms:GenerateDataKey*",
                    "kms:ReEncrypt*",
                  ],
                  "Effect": "Allow",
                  "Resource": "arn:*:kms:*:*:key/*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "AccessPolicy",
          },
        ],
        "RoleName": "KVSStreamCreatorLambdaRole",
      },
      "Type": "AWS::IAM::Role",
    },
    "RegisterDeviceWorkflowKvsCreateLambdaA83CDBA5": {
      "DependsOn": [
        "RegisterDeviceWorkflowKVSStreamCreatorLambdaRole92F0C304",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-123456789012-us-east-1",
          "S3Key": "32d918bda097c43ad86cd13b3d0279caa352c738afc045d18826e367a85f990d.jar",
        },
        "Description": "Lambda responsible for invocation of StepFunction",
        "Environment": {
          "Variables": {
            "airportCode": "iad",
            "tableName": "VLRegisterDeviceJobTable",
          },
        },
        "Handler": "com.amazonaws.videoanalytics.videologistics.workflow.KVSResourceCreateLambda::handleRequest",
        "LoggingConfig": {
          "LogGroup": {
            "Ref": "RegisterDeviceWorkflowDataForwarderLambdaLogGroup2C7FBF4F",
          },
        },
        "MemorySize": 2048,
        "Role": {
          "Fn::GetAtt": [
            "RegisterDeviceWorkflowKVSStreamCreatorLambdaRole92F0C304",
            "Arn",
          ],
        },
        "Runtime": "java17",
        "Timeout": 720,
      },
      "Type": "AWS::Lambda::Function",
    },
    "RegisterDeviceWorkflowVLRegisterDeviceStateMachineA6299E55": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "RegisterDeviceWorkflowVLRegisterDeviceStateMachineRoleDefaultPolicyF76DF5AB",
        "RegisterDeviceWorkflowVLRegisterDeviceStateMachineRoleE55991EF",
      ],
      "Properties": {
        "DefinitionString": {
          "Fn::Join": [
            "",
            [
              "{"StartAt":"Create KVS ARN","States":{"Create KVS ARN":{"Next":"Successful","Retry":[{"ErrorEquals":["Lambda.ClientExecutionTimeoutException","Lambda.ServiceException","Lambda.AWSLambdaException","Lambda.SdkClientException"],"IntervalSeconds":2,"MaxAttempts":6,"BackoffRate":2}],"Catch":[{"ErrorEquals":["States.ALL"],"ResultPath":"$.error","Next":"FailAndCleanupVLRegStatus"}],"Type":"Task","ResultPath":"$.output","Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke","Parameters":{"FunctionName":"",
              {
                "Fn::GetAtt": [
                  "RegisterDeviceWorkflowKvsCreateLambdaA83CDBA5",
                  "Arn",
                ],
              },
              "","Payload":{"JobId.$":"$.partitionKey"}}},"Successful":{"Type":"Succeed"},"FailAndCleanupVLRegStatus":{"Next":"Fail","Retry":[{"ErrorEquals":["Lambda.ClientExecutionTimeoutException","Lambda.ServiceException","Lambda.AWSLambdaException","Lambda.SdkClientException"],"IntervalSeconds":2,"MaxAttempts":6,"BackoffRate":2},{"ErrorEquals":["States.ALL"],"IntervalSeconds":2,"MaxAttempts":5,"BackoffRate":2}],"Type":"Task","ResultPath":"$.output","Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke","Parameters":{"FunctionName":"",
              {
                "Fn::GetAtt": [
                  "RegisterDeviceWorkflowFailAndCleanupVlRegHandler41AB87C8",
                  "Arn",
                ],
              },
              "","Payload":{"jobId.$":"$.partitionKey","FailureReason.$":"$.error.Cause"}}},"Fail":{"Type":"Fail"}}}",
            ],
          ],
        },
        "RoleArn": {
          "Fn::GetAtt": [
            "RegisterDeviceWorkflowVLRegisterDeviceStateMachineRoleE55991EF",
            "Arn",
          ],
        },
      },
      "Type": "AWS::StepFunctions::StateMachine",
      "UpdateReplacePolicy": "Delete",
    },
    "RegisterDeviceWorkflowVLRegisterDeviceStateMachineRoleDefaultPolicyF76DF5AB": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "RegisterDeviceWorkflowKvsCreateLambdaA83CDBA5",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "RegisterDeviceWorkflowKvsCreateLambdaA83CDBA5",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "RegisterDeviceWorkflowFailAndCleanupVlRegHandler41AB87C8",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "RegisterDeviceWorkflowFailAndCleanupVlRegHandler41AB87C8",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "RegisterDeviceWorkflowVLRegisterDeviceStateMachineRoleDefaultPolicyF76DF5AB",
        "Roles": [
          {
            "Ref": "RegisterDeviceWorkflowVLRegisterDeviceStateMachineRoleE55991EF",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "RegisterDeviceWorkflowVLRegisterDeviceStateMachineRoleE55991EF": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "states.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "RegisterDeviceWorkflowWorkflowVLRegisterDeviceJobTableDDBStreamProcessorLambdaRole567D5326": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "RegisterDeviceWorkflowWorkflowVLRegisterDeviceJobTableDDBStreamProcessorLambdaRoleDefaultPolicyF8F7A8E4": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "states:StartExecution",
              "Effect": "Allow",
              "Resource": {
                "Ref": "RegisterDeviceWorkflowVLRegisterDeviceStateMachineA6299E55",
              },
            },
            {
              "Action": [
                "sqs:SendMessage",
                "sqs:GetQueueAttributes",
                "sqs:GetQueueUrl",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "RegisterDeviceWorkflowWorkflowVLRegisterDeviceJobTableVLRegisterDeviceJobTableStreamDLQEB5F14A7",
                  "Arn",
                ],
              },
            },
            {
              "Action": [
                "kms:Decrypt",
                "kms:Encrypt",
                "kms:ReEncrypt*",
                "kms:GenerateDataKey*",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "RegisterDeviceWorkflowWorkflowVLRegisterDeviceJobTableVLRegisterDeviceJobTableStreamDlqEncryptionKeyC8FB4B0C",
                  "Arn",
                ],
              },
            },
            {
              "Action": "dynamodb:ListStreams",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "dynamodb:DescribeStream",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "VLRegisterDeviceJobTableA904F593",
                  "StreamArn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "RegisterDeviceWorkflowWorkflowVLRegisterDeviceJobTableDDBStreamProcessorLambdaRoleDefaultPolicyF8F7A8E4",
        "Roles": [
          {
            "Ref": "RegisterDeviceWorkflowWorkflowVLRegisterDeviceJobTableDDBStreamProcessorLambdaRole567D5326",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "RegisterDeviceWorkflowWorkflowVLRegisterDeviceJobTableVLRegisterDeviceJobTableStreamDLQEB5F14A7": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "KmsMasterKeyId": {
          "Fn::GetAtt": [
            "RegisterDeviceWorkflowWorkflowVLRegisterDeviceJobTableVLRegisterDeviceJobTableStreamDlqEncryptionKeyC8FB4B0C",
            "Arn",
          ],
        },
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "RegisterDeviceWorkflowWorkflowVLRegisterDeviceJobTableVLRegisterDeviceJobTableStreamDlqEncryptionKeyC8FB4B0C": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "EnableKeyRotation": true,
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::123456789012:root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
            {
              "Action": [
                "kms:Create*",
                "kms:Describe*",
                "kms:Enable*",
                "kms:List*",
                "kms:Put*",
                "kms:Update*",
                "kms:Revoke*",
                "kms:Disable*",
                "kms:Get*",
                "kms:Delete*",
                "kms:TagResource",
                "kms:UntagResource",
                "kms:ScheduleKeyDeletion",
                "kms:CancelKeyDeletion",
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::123456789012:root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "RegisterDeviceWorkflowWorkflowVLRegisterDeviceJobTableVLRegisterDeviceJobTableStreamProcessor3042FE4F": {
      "DependsOn": [
        "RegisterDeviceWorkflowWorkflowVLRegisterDeviceJobTableDDBStreamProcessorLambdaRoleDefaultPolicyF8F7A8E4",
        "RegisterDeviceWorkflowWorkflowVLRegisterDeviceJobTableDDBStreamProcessorLambdaRole567D5326",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-123456789012-us-east-1",
          "S3Key": "52046f46f1b49f3050b9ab578eb39a2d45356e865f875bd9b089b807f244b858.jar",
        },
        "Description": "Lambda responsible for invocation of StepFunction",
        "Environment": {
          "Variables": {
            "PARTITION_KEY": "JobId",
            "SORT_KEY": "",
            "STATE_MACHINE_ARN": {
              "Ref": "RegisterDeviceWorkflowVLRegisterDeviceStateMachineA6299E55",
            },
          },
        },
        "Handler": "com.amazonaws.videoanalytics.workflow.lambda.TriggerStepFunctionLambda::handleRequest",
        "MemorySize": 1028,
        "Role": {
          "Fn::GetAtt": [
            "RegisterDeviceWorkflowWorkflowVLRegisterDeviceJobTableDDBStreamProcessorLambdaRole567D5326",
            "Arn",
          ],
        },
        "Runtime": "java17",
      },
      "Type": "AWS::Lambda::Function",
    },
    "RegisterDeviceWorkflowWorkflowVLRegisterDeviceJobTableVLRegisterDeviceJobTableStreamProcessorDynamoDBEventSourceWorkflowStackVLRegisterDeviceJobTable3C9351746FD1294A": {
      "Properties": {
        "BatchSize": 10,
        "BisectBatchOnFunctionError": true,
        "DestinationConfig": {
          "OnFailure": {
            "Destination": {
              "Fn::GetAtt": [
                "RegisterDeviceWorkflowWorkflowVLRegisterDeviceJobTableVLRegisterDeviceJobTableStreamDLQEB5F14A7",
                "Arn",
              ],
            },
          },
        },
        "Enabled": true,
        "EventSourceArn": {
          "Fn::GetAtt": [
            "VLRegisterDeviceJobTableA904F593",
            "StreamArn",
          ],
        },
        "FunctionName": {
          "Ref": "RegisterDeviceWorkflowWorkflowVLRegisterDeviceJobTableVLRegisterDeviceJobTableStreamProcessor3042FE4F",
        },
        "MaximumRetryAttempts": 2,
        "StartingPosition": "LATEST",
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "VLRegisterDeviceJobTableA904F593": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "JobId",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "DeletionProtectionEnabled": true,
        "KeySchema": [
          {
            "AttributeName": "JobId",
            "KeyType": "HASH",
          },
        ],
        "PointInTimeRecoverySpecification": {
          "PointInTimeRecoveryEnabled": true,
        },
        "SSESpecification": {
          "SSEEnabled": true,
        },
        "StreamSpecification": {
          "StreamViewType": "NEW_AND_OLD_IMAGES",
        },
        "TableClass": "STANDARD",
        "TableName": "VLRegisterDeviceJobTable",
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
